async/src/main/java/org/odk/collect/async/WorkerAdapter.kt

```diff
@@ -12,12 +12,16 @@ abstract class WorkerAdapter(
 
     override fun doWork(): Result {
         val stringInputData = inputData.keyValueMap.mapValues { it.value.toString() }
-        val completed = spec.getTask(applicationContext, stringInputData).get()
+        val completed = spec.getTask(applicationContext, stringInputData, isLastUniqueExecution()).get()
 
         return if (completed) {
             Result.success()
-        } else {
+        } else if (spec.maxRetries == null || runAttemptCount < spec.maxRetries!!) {
```

Comment:
I think this should potentially be `runAttemptCount <= spec.maxRetries` (I've left a comment on the test about that). Also, this could avoid `!!` with `spec.maxRetries.let {}` around the `if`.

Reply:
> Also, this could avoid !! with spec.maxRetries.let {} around the if.

Could you share the whole construction. I'm not sure if I can use anything like that here.

Reply:
```kotlin
return spec.maxRetries.let { maxRetries ->
    if (completed) {
        Result.success()
    } else if (maxRetries != null && runAttemptCount < maxRetries) {
        Result.retry()
    } else {
        Result.failure()
    }
}
```

Alternatively, you can just pull out the value:

```kotlin
val maxRetries = spec.maxRetries
return if (completed) {
    Result.success()
} else if (maxRetries == null || runAttemptCount < maxRetries) {
    Result.retry()
} else {
    Result.failure()
}
```